File: python_solutions/1576.replace-all-s-to-avoid-consecutive-repeating-characters.cpp
// C libraries
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>

// Containers
#include <deque>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// Input/Output
#include <fstream>
#include <iomanip>
#include <ios>
#include <iostream>
#include <istream>
#include <ostream>
#include <sstream>

// Other
#include <algorithm>
#include <bitset>
#include <chrono>
#include <iterator>
#include <limits>
#include <random>
#include <stdexcept>
#include <string>
#include <tuple>
#include <type_traits>
#include <utility>

// ==================================================

using namespace std;

// constants
const double EPS = 1e-9;
const int MOD = 1000000007;

// type alias
using pii = pair<int, int>;
using ll = long long;
using ull = unsigned long long;
using vi = vector<int>;
using vb = vector<bool>;
using vc = vector<char>;
using vs = vector<string>;
using vvb = vector<vector<bool>>;
using vvc = vector<vector<char>>;
using vvi = vector<vector<int>>;
using vvs = vector<vector<string>>;
using mii = map<int, int>;
using mci = map<char, int>;
using si = set<int>;
using spii = set<pair<int, int>>;
using umii = unordered_map<int, int>;
using umci = unordered_map<char, int>;
using umsi = unordered_map<string, int>;
using usi = unordered_set<int>;
using usc = unordered_set<char>;
using uss = unordered_set<string>;
using vpii = vector<pair<int, int>>;

typedef struct TreeNode TreeNode;
using ptn = TreeNode *;
using tn = TreeNode;

// ==================================================

// fast IO
static auto __speedup__ = []() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  return 0;
}();

// ==================================================

// some macro for less typing
#define forloop(i, n) for (int i = 0; i < n; i++)             //[0, n)
#define forloopr(i, n) for (int i = n - 1; i >= 0; --i)       // reverse [0, n)
#define forloopup(i, a, b) for (int i = a; i < b; ++i)        // [a, b)
#define forloopdown(i, a, b) for (int i = b - 1; i >= a; --i) // reverse [a, b)
#define forunfold(i, arr) for (auto &i : arr)

#define INF 0x3f3f3f3f
#define MAX(a, b) a = max(a, b)
#define MIN(a, b) a = min(a, b)

#define fi first
#define se second
#define mp make_pair
#define pb push_back
#define pf push_front
#define ef emplace_front
#define eb emplace_back
#define qall(v) v.begin(), v.end()
#define qsize(v) (int)v.size()
#define qsort(v) sort(v.begin(), v.end())
#define rqsort(v) sort(v.rbegin(), v.rend())
#define qreverse(v) reverse(v.begin(), v.end())

// int to string
string itos(int n) { return to_string(n); }
// char to string
string ctos(char c) { return string(1, c); };

inline string upper(string s) {
  string t(s);
  transform(t.begin(), t.end(), t.begin(), ::toupper);
  return t;
}
inline string lower(string s) {
  string t(s);
  transform(t.begin(), t.end(), t.begin(), ::tolower);
  return t;
}

int dirs[5] = {-1, 0, 1, 0, -1};

// ==================================================
/*
 * @lc app=leetcode id=1576 lang=cpp
 *
 * [1576] Replace All ?'s to Avoid Consecutive Repeating Characters
 *
 * https://leetcode.com/problems/replace-all-s-to-avoid-consecutive-repeating-characters/description/
 *
 * algorithms
 * Easy (45.88%)
 * Total Accepted:    6.7K
 * Total Submissions: 14.6K
 * Testcase Example:  '"?zs"'
 *
 * Given a string s containing only lower case English letters and the '?'
 * character, convert all the '?' characters into lower case letters such that
 * the final string does not contain any consecutive repeating characters. You
 * cannot modify the non '?' characters.
 *
 * It is guaranteed that there are no consecutive repeating characters in the
 * given string except for '?'.
 *
 * Return the final string after all the conversions (possibly zero) have been
 * made. If there is more than one solution, return any of them. It can be
 * shown that an answer is always possible with the given constraints.
 *
 *
 * Example 1:
 *
 *
 * Input: s = "?zs"
 * Output: "azs"
 * Explanation: There are 25 solutions for this problem. From "azs" to "yzs",
 * all are valid. Only "z" is an invalid modification as the string will
 * consist of consecutive repeating characters in "zzs".
 *
 * Example 2:
 *
 *
 * Input: s = "ubv?w"
 * Output: "ubvaw"
 * Explanation: There are 24 solutions for this problem. Only "v" and "w" are
 * invalid modifications as the strings will consist of consecutive repeating
 * characters in "ubvvw" and "ubvww".
 *
 *
 * Example 3:
 *
 *
 * Input: s = "j?qg??b"
 * Output: "jaqgacb"
 *
 *
 * Example 4:
 *
 *
 * Input: s = "??yw?ipkj?"
 * Output: "acywaipkja"
 *
 *
 *
 * Constraints:
 *
 *
 *
 * 1 <= s.length <= 100
 *
 *
 * s contains only lower case English letters and '?'.
 *
 *
 *
 */
class Solution {
public:
  string modifyString(string s) {
    char left = 'z';
    for (int i = 0; i < s.size(); i++) {
      if (s[i] == '?') {
        char c = left;
        while (c == left || (i < s.size() - 1 && c == s[i + 1]))
          c = 97 + std::rand() % 26;
        s[i] = c;
      }
      left = s[i];
    }
    return s;
  }
};

auto speed_up = []() {
  ios_base::sync_with_stdio(false);
  return 0;
}();

File: python_solutions/1576.replace-all-s-to-avoid-consecutive-repeating-characters.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007

#
# @lc app=leetcode id=1576 lang=python3
#
# [1576] Replace All ?'s to Avoid Consecutive Repeating Characters
#
# https://leetcode.com/problems/replace-all-s-to-avoid-consecutive-repeating-characters/description/
#
# algorithms
# Easy (45.88%)
# Total Accepted:    6.7K
# Total Submissions: 14.6K
# Testcase Example:  '"?zs"'
#
# Given a string s containing only lower case English letters and the '?'
# character, convert all the '?' characters into lower case letters such that
# the final string does not contain any consecutive repeating characters. You
# cannot modify the non '?' characters.
#
# It is guaranteed that there are no consecutive repeating characters in the
# given string except for '?'.
#
# Return the final string after all the conversions (possibly zero) have been
# made. If there is more than one solution, return any of them. It can be shown
# that an answer is always possible with the given constraints.
#
#
# Example 1:
#
#
# Input: s = "?zs"
# Output: "azs"
# Explanation: There are 25 solutions for this problem. From "azs" to "yzs",
# all are valid. Only "z" is an invalid modification as the string will consist
# of consecutive repeating characters in "zzs".
#
# Example 2:
#
#
# Input: s = "ubv?w"
# Output: "ubvaw"
# Explanation: There are 24 solutions for this problem. Only "v" and "w" are
# invalid modifications as the strings will consist of consecutive repeating
# characters in "ubvvw" and "ubvww".
#
#
# Example 3:
#
#
# Input: s = "j?qg??b"
# Output: "jaqgacb"
#
#
# Example 4:
#
#
# Input: s = "??yw?ipkj?"
# Output: "acywaipkja"
#
#
#
# Constraints:
#
#
#
# 1 <= s.length <= 100
#
#
# s contains only lower case English letters and '?'.
#
#
#
#
import random


class Solution:
    def modifyString(self, s: str) -> str:
        cs = list(s)
        left = 'z'
        for i, c in enumerate(cs):
            if c == '?':
                cur = left
                while cur == left or (i < len(cs) - 1 and cur == cs[i + 1]):
                    cur = chr(random.randint(1, 26) + 97)
                cs[i] = cur
            left = cs[i]
        return ''.join(cs)


sol = Solution()

s = "?zs"
s = "ubv?w"
s = "j?qg??b"
s = "??yw?ipkj?"
print(sol.modifyString(s))
